Updates per feedback (#233)
mulesoft/docs-runtime-fabric: modules/ROOT/pages/upgrade-index.adoc

endif::[]
:page-aliases: upgrade-cluster.adoc, upgrade-components.adoc

> Anypoint Runtime Fabric is a software appliance that combines packages that work together but are upgraded individually:

Reviewer: Wow, this is new. 
Our glossary on the public doc pages has this definition:

Anypoint Runtime Fabric
A container service that automates the configuration, coordination, and management (orchestration), and deployment of your Mule applications and gateways. Runtime Fabric runs on customer-managed infrastructure on AWS, Azure, virtual machines (VMs) or bare-metal servers.

> * Appliance Upgrades

Reviewer: Bullet list items use sentence-style capitalization. They are not substitutes for subtitles or subheads.
```suggestion
* Appliance upgrades
```

+

> ** Appliance upgrades affect platform layer components, which include Docker, Kubernetes, and the Ops Center. These platform layer components are installed on virtual machines (VMs) or bare-metal servers that act as nodes on the cluster.

Reviewer: ```suggestion
** Appliance upgrades affect platform-layer components, which include Docker, Kubernetes, and the Ops Center. These platform-layer components are installed on virtual machines (VMs) or bare-metal servers that act as nodes on the cluster.
```

Reviewer: Is this the only subbullet?
Our style reference recommends not using a list item (or sublist item) if there is only one, which would leave 
* Appliance upgrades
intact but follow it with several plain-text paragraphs (not optimal for list treatment)

+

> Appliance upgrades are applied using the `rtfctl` command-line utility.

Reviewer: Separate paragraph on its own?

Reviewer: Intentionally passive?

+

> Appliance upgrades are applied as a rolling upgrade to each node, one at a time. To minimize impact to running applications, a minimium of 3 controller nodes is required to ensure the cluster remains available. In addition, there should be at least one worker node's worth of unused resources available to allow for applications to be redeployed as each worker node is taken offline to be upgraded.

Reviewer: passive
```suggestion
Appliance upgrades occur on every node but progress from one node to the next, one at a time. To lessen impact to running applications, a minimum of three controller nodes is required to ensure the cluster remains available. In addition, there should be at least one worker node's worth of unused resources available to allow for applications to be redeployed as each worker node is taken offline to be upgraded.
```
What does it mean to have at least three controller nodes? Do we mean that at least three must not be in the process of being upgraded at one time? Doesn't rolling upgrade ensure this?

> * Component Upgrades

Reviewer: ditto prev. edits to first-level bullet above

Reviewer: ditto conversion of subbullet to plain subtext

> ** Runtime Fabric component upgrades involve the internal services that run within the appliance used to facilitate communication to the Anypoint control plane, provide load balancing, and log and metrics forwarding.

Reviewer: "involve the internal services that run within the appliance used to facilitate communication to" is a pretty big mouthful. Can we say simply that upgrades to components include upgrading the control plane, load balancing, and log and metrics forwarding?

+
Runtime Fabric component upgrades are applied within Runtime Manager via a button that's visible when an upgrade is made available.
+
Runtime Fabric component upgrades do not require adding a node.

* `rtfctl` Command-line Utility Upgrades

> ** The `rtfctl` command-line utility is used to manage Runtime Fabric and perform appliance and node configuration upgrades.

Reviewer: ditto prev. edits re. bullet and subbullet presentations

Reviewer: passive. Consider
```suggestion
* `rtfctl` Command-line Utility Upgrades

** The `rtfctl` command-line utility manages Runtime Fabric and performs appliance and node configuration upgrades.
```

+

> Upgrades for the `rtfctl` utility are delivered as a new binary via the *Runtime Fabric downloads* page in Runtime Manager.

Reviewer: Also passive. Is this accurate?
```suggestion
You upgrade the `rtfctl` utility to a new binary by using the *Runtime Fabric downloads* page in Runtime Manager.
```

* Node Configuration Upgrades

> ** Node configuration upgrades involve scripts for provisioning AWS and Azure infrastructure, preparing each node, bootstrapping a new Runtime Fabric, adjusting operating system limits, and performing jobs to clean up system resources.

Reviewer: ditto presentation

Reviewer: ```suggestion
* Node Configuration Upgrades
Node configuration upgrades require you to:
** Provide scripts for provisioning AWS and Azure infrastructure
** Prepare each node <by doing what? do do what?>
** Bootstrap a new instance of Runtime Fabric
** Adjust operating system limits
** Perform jobs to clean up system resources
```
Is there a more specific term for "clean up"? In what way?

+

> Node configuration upgrades are delivered as an updated install script version using the *Runtime Fabric downloads* page in Runtime Manager.

Reviewer: ditto re. rewording to avoid passive

+
[NOTE]

> If you are upgrading the node configuration with updated install scripts, the latest version of the appliance is automatically downloaded and used.

Reviewer: passive

> It is important to run with the most current versions of the appliance and `rtfctl` command-line utility.

Reviewer: ```suggestion
It is important to use the most current versions of both the appliance and the `rtfctl` command-line utility.
```
If you want to use "run," specify what is running.

> == Determine the Component Versions You Are Using

Reviewer: This is a Task topic title for content that is not a task, which requires steps.
GLOBAL

Run the following command on a Runtime Fabric controller node:
```
/opt/anypoint/runtimefabric/rtfctl version
```

Output includes:

> 

Reviewer: I would be tempted to show the output and then call out and define each element, in the same order as presented in actual output.

```

[NOTE]

> The Runtime Fabric component and appliance versions that are in use are also displayed in the *Runtime Fabrics* tab in Runtime Manager. If an upgrade is available, an "Upgrade to vx.x" message is displayed.

Reviewer: passive; reword to tell the audience what it can see by going where?

> == Determine the Most Recent Component Versions Available

Reviewer: Also not a task

The most recent component versions are displayed in the *Downloads* page, located in Runtime Manager in the *Runtime Fabrics* tab. A link to each package’s release notes is provided to show the changes.

You can also review the xref:release-notes::runtime-fabric/runtime-fabric-release-notes.adoc[Anypoint Runtime Fabric Release Notes] information.

== Apply the Upgrades
Before upgrading your production environment, perform the following steps on a nonproduction Runtime Fabric to verify compatibility with your environment.

. Review the xref:release-notes::runtime-fabric/runtime-fabric-release-notes.adoc[Anypoint Runtime Fabric Release Notes] information.

> . Provision additional worker nodes as needed to maintain availability of applications when applying appliance upgrades:

Reviewer: ```suggestion
. Provision additional worker nodes as needed to maintain availability of applications when applying appliance upgrades.
```
Follow with plain text rather than a single substep. GLOBAL

> .. If the amount of available resources for application deployments is less than the resources of one worker node, add another worker node with the same amount of CPU, memory, and disk resources to the Runtime Fabric.

Reviewer: ```suggestion
.. If the amount of available resources for application deployments is less than the resources of one worker node, add another worker node with the same amount of CPU, memory, and disk resources to Runtime Fabric.
```

+
See xref:manage-nodes.adoc[How to Add a Node to Runtime Fabric].
. To upgrade Runtime Fabric components:

.. Plan to upgrade when your team is not deploying new applications or managing existing applications on Runtime Fabric.  
.. Navigate to your Runtime Fabric by selecting it on the *Runtime Fabrics* tab in Runtime Manager.
.. Select "Upgrade to vx.x". The status of the Runtime Fabric is updated to show that the upgrade procedure is in progress.
.. Wait for the status to transition to *Active*. 
+

> In most cases, the upgrade takes less than 10 minutes. When the status transitions to *Active*, and the version of the Runtime Fabric reflects the new version, the component upgrade is complete.

Reviewer: ```suggestion
In most cases, the upgrade takes fewer than 10 minutes. When the status transitions to *Active* and the version shown is the new version, the component upgrade is complete.
```

+
If the Runtime Fabric and Runtime Manager connection disconnects during the upgrade, connectivity is restored after the upgrade finishes.
+
Keep your browser session open to use in later steps.
+
[IMPORTANT]

> To upgrade to Runtime Fabric version 1.5 or later, you must be running version 1.4.54. If you are not running version 1.4.54, you'll be shown a button to upgrade to 1.4.54. Upgrade to this version in order to enable upgrades to 1.5.

Reviewer: ```suggestion
If you see a button offering you the option to upgrade to 1.4.54, you must do so before you can upgrade to Runtime Fabric 1.5 or later. 
```

----
sudo chmod +x rtfctl
----

> . If you are not using the latest version of the appliance, upgrade to the latest version:

Reviewer: ```suggestion
. If you are not using it, upgrade to the latest version of the appliance:
```

.. Make sure the `/opt/anypoint/runtimefabric/installer` directory has a minimum of 4 GiB of disk space available.
.. Verify you have root privileges. Root privileges are required to perform an upgrade.
.. Verify your environment meets the prerequisites:

> *** The internal load balancer must be running on at least 3 replicas to maintain availability. On newer versions of Runtime Fabric, the internal load balancer runs on all controller nodes by default.

Reviewer: ```suggestion
*** The internal load balancer is running on at least three replicas, to maintain availability. 
On newer versions of Runtime Fabric, the internal load balancer runs on all controller nodes by default.
```

> *** An external load balancer must be configured to load balance incoming requests to the controller nodes, with a health check configured for TCP port 443.

Reviewer: ```suggestion
*** An external load balancer is configured to load balance incoming requests to the controller nodes, with a health check configured for TCP port 443.
```

> *** The CPU and memory resources for at least one additional worker node must be available in the cluster.

Reviewer: ```suggestion
*** The CPU and memory resources for at least one additional worker node are available in the cluster.
```

+
This enables safe removal of a worker node from the cluster to apply upgrades without impacting application availability.

> *** Applications serving inbound requests must be scaled to a minimum of 2 replicas.

Reviewer: ```suggestion
*** Applications serving inbound requests are scaled to a minimum of two replicas.
```

Reviewer: The information in this topic is becoming too deeply nested, which is an indication that the topic should be broken up into two or more peers or contain subtopics.
Not something you can do right now, but you should look at this later.

.. Verify Runtime Fabric is healthy:
... Using your web browser, navigate to the *Runtime Fabrics* tab in Runtime Manager and verify that Runtime Fabric reports an *Active* status.
... Open a terminal to a Runtime Fabric controller node.

> ... Run `/opt/anypoint/runtimefabric/rtfctl status` and confirm the cluster status is healthy.

Reviewer: ```suggestion
... Run `/opt/anypoint/runtimefabric/rtfctl status` and confirm that the cluster status is healthy.
```

+
The upgrade can take more than an hour to complete.
+

> The connection between Runtime Fabric and Runtime Manager can disconnect multiple times during the upgrade procedure. Connectivity is restored after the upgrade finishes.

Reviewer: ```suggestion
If the connection between Runtime Fabric and Runtime Manager breaks repeatedly during the upgrade procedure, connectivity is restored after the upgrade finishes.
```

+
. The upgrade procedure streams output to your terminal session. If the previous error reoccurs, perform the troubleshooting steps described in the following section.
 
== Troubleshooting Appliance Upgrade Errors

> A specific sequence of steps is performed during an appliance (installer package) upgrade. If an error occurs, the upgrade pauses and displays an error. In most cases, the availability of running applications is not impacted when running multiple replicas of each application on a production Runtime Fabric configuration.

Reviewer: Intro text is longer and more obvious than it needs to be. Consider something like this:
```suggestion
If an error occurs, the upgrade pauses and displays the reason. In most cases, the availability of multiple running applications on a production Runtime Fabric configuration is not impacted.
```

Most errors are due to insufficient disk performance on the `etcd` block device running on the controller nodes. 

Perform the following steps to resolve common errors:

> . Use the `gravity plan` command to identify the phase where the upgrade paused: 

Reviewer: ```suggestion
. Use the `gravity plan` command to identify where the upgrade paused: 
```

----
. If the upgrade again terminates with an error:
.. Read the logs to identify which node requires repair.

> .. Submit a ticket to MuleSoft Support if additional assistance is required.

Reviewer: passive
```suggestion
.. Submit a ticket to MuleSoft Support if you require additional assistance.
```

If an error is returned, wait a few minutes and repeat the previous steps.

> == Roll Back an Appliance Upgrade

Reviewer: Be sure that this is a task topic, with steps, if using an imperative title.
